### PR TITLE
http: Add tracing

### PR DESCRIPTION
**Pull request by bpon**
Using the [opentracing](http://opentracing.io) API

### CODE REVIEW

File: yarpc-transport-http/src/main/java/io/yarpc/transport/http/TransportRequestDecoderConfiguration.java [hunk 1]
```diff
@@ -66,25 +79,41 @@ public TransportRequest build(
...
-      return now.plusMillis(Long.parseLong(timeout));
+      return start.plusMillis(Long.parseLong(timeout));
     } catch (NumberFormatException e) {
       throw new InvalidTimeoutException(service, procedure, timeout, e);
     }
   }
 
+  private Span createSpan(
+      Instant start, String service, String procedure, String caller, HttpHeaders httpHeaders) {
+    SpanBuilder spanBuilder =
```
Comment (peter-edge) [inline yarpc-transport-http/src/main/java/io/yarpc/transport/http/TransportRequestDecoderConfiguration.java]: nit: wout prefer tracer be in the same line

File: yarpc-transport-http/src/main/java/io/yarpc/transport/http/TransportRequestDecoderConfiguration.java [hunk 2]
```diff
@@ -66,25 +79,41 @@ public TransportRequest build(
...
+        tracer
+            .buildSpan(procedure)
+            .withStartTimestamp(Instant.EPOCH.until(start, ChronoUnit.MICROS))
+            .withTag(SpanTags.CALLER, caller)
+            .withTag(SpanTags.SERVICE, service)
+            .withTag(SpanTags.ENCODING, httpHeaders.get(HeaderMapper.ENCODING))
+            .withTag(SpanTags.TRANSPORT, "http");
```
Comment (peter-edge) [inline yarpc-transport-http/src/main/java/io/yarpc/transport/http/TransportRequestDecoderConfiguration.java]: nit: http as private static final?

File: yarpc-transport-http/src/main/java/io/yarpc/transport/http/TransportRequestDecoderConfiguration.java [hunk 3]
```diff
@@ -66,25 +79,41 @@ public TransportRequest build(
...
+    SpanContext parent =
```
Comment (peter-edge) [inline yarpc-transport-http/src/main/java/io/yarpc/transport/http/TransportRequestDecoderConfiguration.java]: same line, newlines after each parameter?

File: yarpc-transport-http/src/test/java/io/yarpc/transport/http/TransportRequestDecoderTest.java [hunk 4]
```diff
@@ -69,7 +71,8 @@ public void before() {
               resultTester = req.getBody().accumulate().test();
               return invocation.getMock();
             });
-    decoder = new HttpTransportDecoder<>(TransportRequestDecoder.CONFIGURATION);
+    Tracer tracer = NoopTracerFactory.create();
+    decoder = new HttpTransportDecoder<>(new TransportRequestDecoderConfiguration(tracer));
```
Comment (peter-edge) [inline yarpc-transport-http/src/test/java/io/yarpc/transport/http/TransportRequestDecoderTest.java]: nit: just put `NoopTracerFactory.create()` in the `TransportRequestDecoderConfiguration` constructor, this is similar to other parts of the code here

File: yarpc-transport-http/src/test/java/io/yarpc/transport/http/TransportRequestEncoderTest.java [hunk 5]
```diff
@@ -56,14 +59,16 @@
...
   @Before
   public void before() throws Exception {
-    encoder =
-        new HttpTransportEncoder<>(
-            new TransportRequestEncoderConfiguration(
-                new URL("http://test.uber.internal:12345/foo")));
+    Tracer tracer = NoopTracerFactory.create();
+    encoderConfig =
+        new TransportRequestEncoderConfiguration(
+            new URL("http://test.uber.internal:12345/foo"), tracer);
```
Comment (peter-edge) [inline yarpc-transport-http/src/test/java/io/yarpc/transport/http/TransportRequestEncoderTest.java]: nit: same

File: yarpc-transport-http/src/main/java/io/yarpc/transport/http/HttpTransport.java [hunk 6]
```diff
@@ -48,6 +51,8 @@
   @NonNull @Getter private final NioEventLoopGroup workerGroup = new NioEventLoopGroup();
   @NonNull @Getter private final NioEventLoopGroup bossGroup = new NioEventLoopGroup();
 
+  @NonNull @Getter private final Tracer tracer;
```
Comment (bpon) [inline yarpc-transport-http/src/main/java/io/yarpc/transport/http/HttpTransport.java]: Sure.  The only downside is that it introduces a dependency on opentracing-noop, but it has no external dependencies, so not a big deal.

File: yarpc-transport-http/src/main/java/io/yarpc/transport/http/TransportRequestDecoderConfiguration.java (hunk 3 above)
Comment (bpon) [inline yarpc-transport-http/src/main/java/io/yarpc/transport/http/TransportRequestDecoderConfiguration.java]: I would like to, but the auto-formatter does this.

File: yarpc-transport-http/src/main/java/io/yarpc/transport/http/TransportRequestDecoderConfiguration.java (hunk 2 above)
Comment (bpon) [inline yarpc-transport-http/src/main/java/io/yarpc/transport/http/TransportRequestDecoderConfiguration.java]: I don't think that making this a private static final improves readability in this case.

I would really like for this to be `HttpTransport.NAME`, but for some reason we use lowercase here and uppercase everywhere else (which is consistent with yarpc-go).  We could also do `HttpTransport.NAME.toLowerCase()` but the `toLowerCase()` part seems kind of arbitrary, so I'd prefer to leave this as is.

**Pull request merged**